git/filter_protocol.go

```diff
@@ -0,0 +1,132 @@
+package git
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strconv"
+	"strings"
+)
+
+type protocol struct {
+	r *bufio.Reader
+	w *bufio.Writer
+}
+
+func newProtocolRW(r io.Reader, w io.Writer) *protocol {
+	return &protocol{
+		r: bufio.NewReader(r),
+		w: bufio.NewWriter(w),
+	}
+}
+
+// readPacket reads a single packet entirely and returns the data encoded within
+// it. Errors can occur in several cases, as described below.
+//
+// 1) If no data was present in the reader, and no more data could be read (the
+//    pipe was closed, etc) than an io.EOF will be returned.
+// 2) If there was some data to be read, but the pipe or reader was closed
+//    before an entire packet (or header) could be ingested, an
+//    io.ErrShortBuffer error will be returned.
+// 3) If there was a valid header, but no body associated with the packet, an
+//    "Invalid packet length." error will be returned.
+// 4) If the data in the header could not be parsed as a hexadecimal length in
+//    the Git pktline format, the parse error will be returned.
+//
+// If none of the above cases fit the state of the data on the wire, the packet
+// is returned along with a nil error.
+func (p *protocol) readPacket() ([]byte, error) {
+	var pktLenHex [4]byte
+	if n, err := io.ReadFull(p.r, pktLenHex[:]); err != nil {
```

Comment:
I think this should have an explicit timeout on it since this is a separate process and could be orphaned if the git process terminated unexpectedly on an OS that doesn't automatically terminate child processes (e.g. Windows). 

You could still use `ReadFull` and just use a signal channel e.g. 
```go
	var n int
	var err error
	sig := make(chan int, 1)
	go func() {
		n, err = io.ReadFull(p.r, pktLenHex[:])
		sig <- 1
	}()
	select {
	case <-sig:
		break
	case <-time.After(time.Second * 30):
		return nil, fmt.Errorf("Timeout reading from git")
	}
	// n & err OK to process
```